cooked-validators/src/Cooked/Tx/Constraints.hs

```diff
@@ -27,78 +29,132 @@ import qualified PlutusTx as Pl
+          <> Pl.unspentOutputs (M.singleton oref o)
+      constr = Pl.mustSpendScriptOutput oref (Pl.Redeemer $ Pl.toBuiltinData r)
+  toLedgerConstraint (SpendsPK (oref, o)) = (lkups, constr)
+    where
+      lkups = Pl.unspentOutputs (M.singleton oref o)
+      constr = Pl.mustSpendPubKeyOutput oref
+  toLedgerConstraint (Mints Nothing pols v) = (lkups, constr)
+    where
+      lkups = foldMap Pl.mintingPolicy pols
+      constr = Pl.mustMintValue v
+  toLedgerConstraint (Mints (Just r) pols v) = (lkups, constr)
+    where
+      lkups = foldMap Pl.mintingPolicy pols
+      constr = Pl.mustMintValueWithRedeemer (Pl.Redeemer (Pl.toBuiltinData r)) v
+  toLedgerConstraint (Before t) = (mempty, constr)
+    where
+      constr = Pl.mustValidateIn (Pl.to t)
+  toLedgerConstraint (After t) = (mempty, constr)
+    where
+      constr = Pl.mustValidateIn (Pl.from t)
+  toLedgerConstraint (ValidateIn r) = (mempty, Pl.mustValidateIn r)
+  toLedgerConstraint (SignedBy hashes) = (mempty, foldMap (Pl.mustBeSignedBy . Pl.PaymentPubKeyHash) hashes)
+
+instance Constraint OutConstraint where
+  extractDatumStr (PaysScript _validator datumsAndValues) =
+    M.fromList
+      . map ((\d -> (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData d, show d)) . fst)
+      $ datumsAndValues
+  extractDatumStr (PaysPKWithDatum _pk _stak mdat _v) =
+    maybe M.empty (\d -> M.singleton (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData d) (show d)) mdat
+
+  toLedgerConstraint (PaysPKWithDatum p stak dat v) = (lkups, constr)
+    where
+      mData = fmap (Pl.Datum . Pl.toBuiltinData) dat
+
+      lkups =
+        maybe mempty Pl.otherData mData
+          -- TODO: do we want to akk ownStakePubKeyHash on 'PaysPKWithDatum'? Would we rather have
+          -- a different 'WithOwnStakePubKeyHash' constraint?
+          <> maybe mempty Pl.ownStakePubKeyHash stak
+      constr = Pl.singleton $ Pl.MustPayToPubKeyAddress (Pl.PaymentPubKeyHash p) stak mData v
+  toLedgerConstraint (PaysScript v outs) = (lkups, constr)
+    where
+      lkups = Pl.otherScript (Pl.validatorScript v)
+      constr = mconcat $
+        flip map outs $ \(d, val) ->
+          Pl.mustPayToOtherScript
+            (Pl.validatorHash $ Pl.validatorScript v)
+            (Pl.Datum $ Pl.toBuiltinData d)
+            val
+
+instance Constraint Constraints where
+  extractDatumStr (miscConstraints :=>: outConstraints) =
+    M.union
+      (M.unions (extractDatumStr <$> miscConstraints))
+      (M.unions (extractDatumStr <$> outConstraints))
+
+  toLedgerConstraint (miscConstraints :=>: outConstraints) =
+    (mconcat lkups, mconcat constrs)
+    where
+      (lkups, constrs) =
+        unzip $
+          (toLedgerConstraint <$> miscConstraints)
+            <> (toLedgerConstraint <$> outConstraints)
+
+-- | Look for the 'Pl.TxOut' corresponding to an output constraint
+-- 'OutConstraint' among the outputs of a transaction.
+findGeneratedTxOutput :: OutConstraint -> [Pl.TxOut] -> Maybe Pl.TxOut
```

Comment:
Isn't it easier to map an `OutContraint` to a `Pl.TxOut`, then define `orderTxOutputs` as:

```haskell
orderTxOutputs :: [OutConstraint] -> [Pl.TxOut] -> [Pl.TxOut]
orderTxOutputs os _ = map outContrsintToTxOut os
```

There's always the case where the given list of `[Pl.TxOut]` might contain something else; we can either return a `Maybe` or just explicitely add it there:

```haskell
orderTxOutputs :: [OutConstraint] -> [Pl.TxOut] -> [Pl.TxOut]
orderTxOutputs expctd given =
  let res = map outContrsintToTxOut expctd
   in res ++ (given Data.List.\\ res)
```

Reply:
Yes it is. Finding the txOuts means generating the txOuts. This follows our discussion about the previous PR on generating everything ourselves and the risks of not doing it exactly like plutus. Here it fails explicitly if we happen not to generate it exactly as Plutus does. We do not risk to add new transaction at the beginning without removing their counterpart later from the list in case of slight mismatch.

Eventually, once we are certain about the generation process (especially the staking pkh), I agree with you.

Reply:
But I think that in this case, its very easy to map `OutConstraint` to `Pl.TxOut`, there's no need to worry about staking:

```haskell
outConstraintToTxOut :: OutConstraint -> Pl.TxOut
outConstraintToTxOut (PaysPKWithDatum pk mdat mstak val) = 
   TxOut { txoutAddress = PubKeyAddress (PubKeyCredential pk) mstak
         , txoutDatumHash = fmap datumHash mdat
         , txoutValue = val
         }
```

Reply:
Done :+1: 
I used `Pl.scriptAddress` to generate the address from `PaysScript`.